cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1158,7 +1179,7 @@ class ObjectInputFile final : public io::RandomAccessFile {
     req.SetKey(ToAwsString(path_.key));
 
     ARROW_ASSIGN_OR_RAISE(auto client_lock, holder_->Lock());
-    auto outcome = client_lock->HeadObject(req);
+    auto outcome = std::move(client_lock)->HeadObject(req);
```

Comment:
TBH I'm not sure if calling `std::move` here is enough to destroy `client_lock` at the end of this statement. @bkietz Can you advise?

Reply:
`std::move` will not destroy this lock (in general, it's useful to think of `std::move` as doing nothing except flagging a reference as legal to move construct from). What you want here is
```suggestion
    S3Model::HeadObjectOutcome outcome;
    {
      ARROW_ASSIGN_OR_RAISE(auto client_lock, holder_->Lock());
      outcome = std::move(client_lock)->HeadObject(req);
    } // client lock goes out of scope here
```

or maybe
```suggestion
    ARROW_ASSIGN_OR_RAISE(auto outcome, holder_->Lock().Map([&](auto client_lock) {
      return client_lock->HeadObject(req);
    }));
```

Reply:
Or you might even prefer
```c++
template <auto MemberPtr, typename... Args>
auto S3ClientHolder::LockThenCall(Args&&... args) {
  return Lock().Map([&](S3ClientLock client_lock) {
    return (client_lock.get()->*MemberPtr)(std::forward<Args>(args)...);
  });
}
```

used like
```suggestion
    ARROW_ASSIGN_OR_RAISE(auto outcome, holder_->LockThenCall<&S3Client::HeadObject>(req));
```

Reply:
Thanks @bkietz !

Reply:
Ok, I opted for a more manual approach because mixing `arrow::Result` and `Aws::Outcome` is not really convenient:
https://github.com/apache/arrow/pull/36536/files#diff-27b767cd06b879c40106ea98d21d3b12f13ec6e9f9aa49218d10583c5c1899c4R725-R726